walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,120 @@
+    long mtngDuration = request.getDuration();
+    Collection<String> attendees = request.getAttendees();
+    Collection<Event> allEvents = events;
+
+    HashMap<Integer, Integer> mandatoryEvents = makeMandatoryEventMap(allEvents, attendees);
+    Collection<TimeRange> meetingTimes = findMeetingTimes(mtngDuration, mandatoryEvents);
+    return meetingTimes;
+  }
+
+
+  /** 
+  * Create a hash map <event start time, event end time> that ONLY contains meetings where 1 or more people from attendees must attend. 
+  **/
+  public HashMap<Integer, Integer> makeMandatoryEventMap (Collection<Event> allEvents, Collection<String> attendees) {
+    HashMap<Integer, Integer> mandatoryEvents = new HashMap<Integer, Integer>();
+
+    for (Event event : allEvents) {
+      //people who must attend 
+      Set<String> people = event.getAttendees();
+      //check if any people in meeting request is present in this set
+      for (String person : attendees) {
+        if (people.contains(person)) {
+          // If there's already an event with this start time in the map, but the current end time is later, add the current instead.  
+          int eventStartTime = event.getWhen().start();
+          int eventEndTime = event.getWhen().end();
+          if (mandatoryEvents.containsKey(eventStartTime)) {
+            if (eventEndTime > mandatoryEvents.get(eventStartTime)) {
+              mandatoryEvents.put(eventStartTime, eventEndTime);
+            }
+            else {
+              break;
+            }
+          }
+          else {
+            mandatoryEvents.put(eventStartTime, eventEndTime);
+          }
+        }  
+      }
+    }
+
+    return mandatoryEvents;
+  }
+
+  /**
+  * Create and return an ArrayList of TimeRange objects that span available meeting times.
+  */
+  public Collection<TimeRange> findMeetingTimes (long mtngDuration, HashMap<Integer, Integer> mandatoryEvents) {
+    Collection<TimeRange> meetingTimes = new ArrayList<TimeRange>();
+    int currTime = TimeRange.START_OF_DAY;
+    int startTime = TimeRange.START_OF_DAY;
+    int currDuration = 0;
+
+    while (currTime < TimeRange.END_OF_DAY) {
+      // If there's an event starting at this time, compare the current duration and meeting duration to see if there is a possible meeting time. 
+      // Reset the current duration variable, move the current time to the end time of the event, and repeat loop. 
+      if (mandatoryEvents.containsKey(currTime)) {
+        int longestMeetingEnd = mandatoryEvents.get(currTime);
+        if (currDuration >= mtngDuration) {
+          // Not inclusive because a meeting starts at this time 
+          meetingTimes.add(TimeRange.fromStartEnd(startTime, currTime, /* inclusive = */ false));
+        }
+        currDuration = 0;
+
+        // Check if there is a conflict: another meeting that starts during this one, and goes later
+        longestMeetingEnd = isConflict(currTime, longestMeetingEnd, mandatoryEvents);
```

Comment:
Methods with names like isFoo are expected to return booleans.  Instead, since this returns an int, it should be something like getLastEventEnd or something that better describes what int you're getting here.